draft-ietf-dprive-dnsoquic.md

```diff
@@ -394,12 +394,17 @@ A client MAY take advantage of the connection resume mechanisms supported by
 QUIC transport {{!RFC9000}} and QUIC TLS {{!RFC9001}}. Clients SHOULD consider
 potential privacy issues associated with session resume before deciding to use
 this mechanism. These privacy issues are detailed in
-{{privacy-issues-with-session-resume}}.
+{{privacy-issues-with-session-resume}} and {{privacy-issues-with-0-rtt-data}},
+and the implementation considerations are discussed in
+{{using-0-rtt-and-resumption}}.
 
-When resuming a session, a client MAY take advantage of the 0-RTT mechanism
-supported by QUIC. The 0-RTT mechanism MUST NOT be used to send data that is
+The 0-RTT mechanism MUST NOT be used to send data that is
 not "replayable" transactions. For example, a client MAY transmit a Query as
```

Comment:
@saradickinson I would like your advice here. The DNS operations are defined in an IANA registry that we could quote -- https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-5. The registry currently defines the following:

OpCode | Name | Reference
-- | -- | --
0 | Query | [RFC1035]
1 | IQuery  (Inverse Query, OBSOLETE) | [RFC3425]
2 | Status | [RFC1035]
3 | Unassigned |  
4 | Notify | [RFC1996]
5 | Update | [RFC2136]
6 | DNS Stateful Operations (DSO) | [RFC8490]
7-15 | Unassigned |  

Of all of these, I see the benefit of allowing `Query` in 0-RTT, and the danger of allowing `Update` or `DSO`. I am not sure about `IQuery`, `Status` and `Notify`. Advice?

Reply:
Since `IQuery` is obsolete, I don't think there is any danger (generates a NotImp AFAIK). `Status` is similar and there was an effort 2 years ago to deprecate it but it failed because... well, DNSOP :-(  Servers respond with one of NotImp, Refused or give no response at all.
`Notify` is worth thinking about. Since it only effects secondaries and it can already be abused to trigger unnecessary SOA or XFR requests to a primary, most implementations already throttle the speed at which they will send those requests in response to one or more `Notify` messages. Some operators require `Notifies` to be TSIG signed before acting on the (which is also true of `Update`). TSIG does provide a defence against replay but note that RFC8945 recommends a default Fudge value (i.e. time offset for validation) of 300s.

Reply:
So I think the text should say that only query with operation type "Query" should be considered in 0-RTT. Is there a case where some servers should *not* allow some specific queries, e.g., because they are some kind of "active" query that starts a script?

Reply:
This is what https://datatracker.ietf.org/doc/html/draft-ietf-dprive-early-data-00 proposed, but this was questioned in the following discussion that then petered out... (an IANA registry of allowed QTYPES was proposed). I notice that that doc was adopted by DPRIVE but then seemed to stall and has now expired.

Reply:
I would rather not quote an expired draft. I also agree with Martin's general recommendation to use application level response codes rather than QUIC errors. Whenever we create a QUIC error, implementation have to end up mapping the QUIC error in some kind of transport semantic. How about giving servers the following choices:

1) Process the request, if the operation type is "QUERY" and the query can be processed without side effect,
2) Defer processing until the connection has been confirmed by the client (quote RFC 9001),
3) Reject the request and prepare a DNS REFUSED error.

If the WG finally produces a 0-RTT draft, it would update DoQ.